docs/rfc/rfc-002-ipc-ecosystem.md

```diff
@@ -0,0 +1,348 @@
+(genesis data, transactions, commits), node status (network info, health
+checks), application state (abci_query, abci_info), mempool state, and other
+attributes of the node and its application. The service also provides methods
+allowing transactions and evidence to be injected ("broadcast") into the
+blockchain.
+
+The RPC service is exposed in several ways:
+
+- HTTP GET: Queries may be sent as URI parameters, with method names in the path.
+
+- HTTP POST: Queries may be sent as JSON-RPC request messages in the body of an
+  HTTP POST request.  The server uses a custom implementation of JSON-RPC that
+  is not fully compatible with the [JSON-RPC 2.0 spec][json-rpc], but handles
+  the common cases.
+
+- Websocket: Queries may be sent as JSON-RPC request messages via a websocket.
+  This transport uses more or less the same JSON-RPC plumbing as the HTTP POST
+  handler.
+
+  TODO: Three methods, maybe related to event subscription, are _only_ exported
+  via websocket. I think these may be used by the light client for
+  synchronization. In any case, the light client appears to be the only direct
+  consumer of the websocket transport at least within the core repository.
+  Figure out whether that's true.
+
+- gRPC: A subset of queries may be issued in protocol buffer format to the gRPC
+  interface described above under (4). As noted, this endpoint is deprecated
+  and will be removed in v0.36.
+
+### Opportunities for Simplification
+
+**Claim:** There are too many IPC mechanisms.
+
+The preponderance of ABCI usage is via the Cosmos SDK, which means the
+application and the consensus node are compiled together into a single binary,
+and the consensus node calls the ABCI methods of the application directly as Go
+functions.
+
+TODO: What are the remaining use cases for the socket protocol and gRPC? Ethan
+mentioned some cases like web-based explorer UIs and possibly other tools.
+
+As far as I can tell the primary consumers of the multi-headed "RPC service"
+are the light client and the `tendermint` command-line client. There is
+probably some local use via curl, but I expect that is mostly ad hoc. Ethan
+reports that nodes are often configured with the ports to the RPC service
+blocked, which is good for security but complicates use by the light client.
+
+TODO: Figure out what is going on with the weird pseudo-inheritance service
+interface.  The BaseService type makes sense, but its embeds are more
+complicated.
+
+### Context: Remote Signer Issues
+
+Since the remote signer needs a secure communication channel to exchange keys
+and signatures, and is expected to run truly remotely from the node (i.e., on a
+separate physical server), there is not a whole lot we can do here. We should
+finish the deprecation and removal of the "raw" socket protocol between the
+consensus node and remote signers, but the use of gRPC is appropriate.
+
+The main improvement we can make is to simplify the implementation quite a bit,
```

Comment:
![](https://media0.giphy.com/media/KzDqC8LvVC4lshCcGK/giphy.gif)

Reply:
Haha, but are there specific details you have in mind that could be simplified?

Reply:
Yes, but I didn't want to go too far down the "implementation details" rabbit hole in this doc—that's more of an ADR. There's an extra layer of abstraction to handle "raw" vs. gRPC that could be ripped out, and some plumbing related to which side is the client vs. the server.